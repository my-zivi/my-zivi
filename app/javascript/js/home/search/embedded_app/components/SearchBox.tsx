import React, { forwardRef, useEffect, useImperativeHandle, useMemo, useRef, useState } from 'preact/compat';
import { delay } from 'lodash';
import { debounceDelay } from 'js/home/search/embedded_app/helpers/SearchBoxHelper';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const searchIcon = require('../../../../../images/home/search.svg');

interface Props {
  currentRefinement: string;

  refine(value?: string): void;

  onFocus: () => void;
  onBlur: () => void;
}

export type SearchBoxRef = {
  autocompleteSearch: (string) => void,
};

const SearchBox = forwardRef<SearchBoxRef, Props>(({ currentRefinement, refine, onFocus, onBlur }, ref) => {
  const [currentInputValue, setCurrentInputValue] = useState(currentRefinement);
  const [debounceTimerId, setDebounceTimerId] = useState(null);
  const currentDelay = useMemo(() => debounceDelay(), []);
  const inputRef = useRef<HTMLInputElement>();

  const debouncedSearch = (value) => {
    if (debounceTimerId) {
      clearInterval(debounceTimerId);
    }

    setDebounceTimerId(delay(refine, currentDelay(), value));
  };

  useImperativeHandle(ref, () => ({
    autocompleteSearch: (value) => {
      setCurrentInputValue(value);
      refine(value);
      inputRef.current.blur();
    },
  }));

  useEffect(() => {
    const term = currentInputValue.trim();
    if (term !== currentRefinement && (term.length > 1 || term.length === 0)) {
      debouncedSearch(term);
    }
  }, [currentInputValue]);

  return (
    <div className="search-field">
      <input
        ref={inputRef}
        type="search"
        value={currentInputValue}
        placeholder={MyZivi.translations.search.searchPlaceholder}
        onFocus={onFocus}
        onBlur={onBlur}
        onChange={(e) => setCurrentInputValue(e.currentTarget.value)}
        onKeyUp={
          (e) => {
            if (e.key.toLowerCase() === 'enter') {
              inputRef.current.blur();
            }
          }
        }
      />

      <button className="search-button" onClick={() => inputRef.current.blur()}>
        <img src={searchIcon} alt={MyZivi.translations.search.search} className="w-100 h-100" />
      </button>
    </div>
  );
});

SearchBox.displayName = 'SearchBox';
export default SearchBox;
