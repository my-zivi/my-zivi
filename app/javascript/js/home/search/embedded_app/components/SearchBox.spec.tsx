import { shallow } from 'enzyme';
import React from 'preact/compat';
import SearchBox, { SearchBoxRef } from './SearchBox';

const noop = () => null;

describe('SearchBox', () => {
  it('renders correctly', () => {
    expect(
      shallow(<SearchBox currentRefinement="" refine={noop} onBlur={noop} onFocus={noop} />),
    ).toMatchSnapshot();
  });

  it('accepts imperative handle', () => {
    const ref = React.createRef<SearchBoxRef>();
    const refine = jest.fn();
    const rendered = shallow(<SearchBox ref={ref} currentRefinement="" refine={refine} onBlur={noop} onFocus={noop} />);
    ref.current.autocompleteSearch('my-search');
    expect(refine).toHaveBeenCalledWith('my-search');

    expect(rendered.update().find('.search-field input').props()).toHaveProperty('value', 'my-search');
  });
});
