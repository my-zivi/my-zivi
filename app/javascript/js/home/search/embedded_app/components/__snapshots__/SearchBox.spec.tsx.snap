// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders correctly 1`] = `
<div
  className="search-field"
>
  <input
    onKeyUp={[Function]}
    onfocusin={[Function]}
    onfocusout={[Function]}
    oninput={[Function]}
    placeholder="string"
    type="search"
    value=""
  />
  <button
    className="search-button"
    onClick={[Function]}
  >
    <img
      alt="string"
      className="w-100 h-100"
      src="test-file-stub"
    />
  </button>
</div>
`;
